soroban-sdk/src/cost_estimate.rs

```diff
@@ -0,0 +1,111 @@
+    /// production scenarios (e.g. certain XDR rountrips). This also doesn't try
+    /// to model resources related to the transaction size.
+    ///
+    /// The returned value is as useful as the preceding setup, e.g. if a test
+    /// contract is used instead of a Wasm contract, all the costs related to
+    /// VM instantiation and execution, as well as Wasm reads/rent bumps will be
+    /// missed.    
+    pub fn resources(&self) -> Option<InvocationResources> {
+        self.env.host().get_last_invocation_resources()
+    }
+
+    /// Estimates the fee for the last invocation's resources, i.e. the
+    /// resources returned by `resources()`.
+    ///
+    /// In order to get non-`None` results, `enable()` has to
+    /// be called and at least one invocation has to happen after that.
+    ///
+    /// The fees are computed using the snapshot of the Stellar Pubnet fees made
+    /// on 2024-12-11.
+    ///
+    /// Take the return value with a grain of salt as both the resource estimate
+    /// and the fee rates may be imprecise.
+    ///
+    /// The returned value is as useful as the preceding setup, e.g. if a test
+    /// contract is used instead of a Wasm contract, all the costs related to
+    /// VM instantiation and execution, as well as Wasm reads/rent bumps will be
+    /// missed.    
+    pub fn fee(&self) -> Option<FeeEstimate> {
+        // This is a snapshot of the fees as of 2024-12-11.
+        let pubnet_fee_config = FeeConfiguration {
+            fee_per_instruction_increment: 25,
+            fee_per_read_entry: 6250,
+            fee_per_write_entry: 10000,
+            fee_per_read_1kb: 1786,
+            // This is a bit higher than the current network fee, it's an
+            // overestimate for the sake of providing a bit more conservative
+            // results in case if the state grows.
+            fee_per_write_1kb: 12000,
+            fee_per_historical_1kb: 16235,
+            fee_per_contract_event_1kb: 10000,
+            fee_per_transaction_size_1kb: 1624,
+        };
+        let pubnet_persistent_rent_rate_denominator = 2103;
+        let pubnet_temp_rent_rate_denominator = 4206;
+        if let Some(resources) = self.resources() {
+            Some(resources.estimate_fees(
+                &pubnet_fee_config,
+                pubnet_persistent_rent_rate_denominator,
+                pubnet_temp_rent_rate_denominator,
+            ))
+        } else {
+            None
+        }
+    }
+
+    /// Returns the detailed CPU and memory metering information recorded thus
+    /// far.
+    ///
+    /// The metering resets before every top-level contract level invocation.
+    ///
+    /// Note, that unlike `resources()`/`fee()` this will always return some
+    /// value. If there was no contract call, then the resulting value will
+    /// correspond to metering any environment setup that has been made thus
+    /// far.
+    pub fn detailed_metering(&self) -> Budget {
```

Comment:
Is this necessary since there is already an `Env::budget()` function

Reply:
`Env::budget()` can be deprecated in favor of more consistent interface.

Reply:
Got it. The budget contains metering info as well as a bunch of auxiliary things (such as the cost schedule and trackers) that the user won't need (and might be confused to equate it with "detailed_metering"), and you can call methods on it. 
For the sake of UX, instead of returning the full `Budget`, it may be better to create a read-only "detailed_metering" type that just contains cpu and mem consumption for various costs, similar what's being shown by Budget::Display.

Reply:
Maybe it could be better if we designed this from scratch, but currently this would just duplicate the functionality that we already have in budget. Maybe we just should leave budget as is and not try to shove it into the new class, as I agree that it seems kind of odd to have the same exact thing in two different places (and we can't easily get rid of budget either as it would be a breaking change). WDYT @leighmcculloch ?

Reply:
+1 we can deprecate Env::budget() in this PR and say to use the new interface. It's helpful to group these together into a single location to look at this kind of stuff.

Reply:
Ok, I've added a deprecation notice. Still not sure about the naming. Maybe `budget` is more appropriate, as it does have some budget-related functionality. I don't have a strong preference here though.